src/OrchardCore.Modules/OrchardCore.AutoSetup/Extensions/DistributedLockExtensions.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Threading.Tasks;
+using OrchardCore.Locking;
+using OrchardCore.Locking.Distributed;
+using OrchardCore.AutoSetup.Options;
+
+namespace OrchardCore.AutoSetup.Extensions
+{
+    /// <summary>
+    /// The distributed lock extensions for Auto Setup.
+    /// </summary>
+    public static class DistributedLockExtensions
+    {
+        /// <summary>
+        /// Lock Prefix
+        /// </summary>
+        private const string LockPrefix = "AUTOSETUP_LOCK";
+
+        /// <summary>
+        /// The default timeout in milliseconds to acquire a distributed setup lock.
+        /// </summary>
+        private const int DefaultLockTimeout = 20_000;
+
+        /// <summary>
+        /// The default expiration in milliseconds of the distributed setup lock.
+        /// </summary>
+        private const int DefaultLockExpiration = 20_000;
+
+        /// <summary>
+        /// Tries to acquire a setup lock 
+        /// </summary>
+        /// <param name="distributedLock">
+        /// The distributed Lock.
+        /// </param>
+        /// <param name="shellName">
+        /// The shell Name for which lock acquires.
+        /// </param>
+        /// <param name="lockOptions">
+        /// The setup options.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Task"/>.
+        /// </returns>
+        public static Task<(ILocker locker, bool locked)> TryAcquireAutoSetupLockAsync(this IDistributedLock distributedLock, string shellName, LockOptions lockOptions)
+        {
+            if (lockOptions == null)
+            {
+                throw new ArgumentNullException(nameof(lockOptions));
+            }
+
+            return distributedLock.TryAcquireLockAsync(
+                $"{LockPrefix}_{lockOptions.LockName ?? shellName}",
```

Comment:
So here we don't need any prefix, configurable LockName, and don't need the shell name, we only need a lock key e.g. "AUTOSETUP_LOCK", because each tenant will be setup by itself on demand, the tenant name will be used as a prefix automatically by the RedisLock.

And about sharing the same Redis Instance by different applications, as said it is another concern, would be good to first try it with only one set of OC instances related to the same database.

Then if you want to try it with 2 sets of OC instances sharing the same Redis instance, if it is easier for you maybe i will merge #9522, i think it is safe to be merged because it only adds a prefix that is null by default